File: llmrouter/src/llmrouter/backends.py
"""Backend registry, live health signals, and the circuit breaker.

Axis 2 of the design: *where* the model runs. The literature almost never models
this, which is why the router scores (model, backend) PAIRS rather than models.

Two operational facts that shape this module:
  * LLM serving pods take **5–15 minutes** to become ready (pull weights, load to
    VRAM, warm KV cache), so reactive autoscaling is useless. Route AROUND cold
    capacity rather than waiting for it.
  * A 10× traffic spike that saturates a local GPU cascades every request to the
    cloud at ~100× the cost. Local timeouts must be aggressive, not 30s.
"""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Mapping, Sequence

from .types import DeploymentClass

#: Aggressive local timeout. The cascade trap: a saturated GPU times everything
#: out, and everything then lands on the expensive path at once.
DEFAULT_LOCAL_TIMEOUT_S = 8.0
DEFAULT_API_TIMEOUT_S = 60.0


@dataclass(frozen=True)
class BackendSpec:
    backend_id: str
    deployment: DeploymentClass = DeploymentClass.API
    models: tuple[str, ...] = ()
    base_url: str | None = None
    residency_tags: frozenset[str] = frozenset()
    priority: int = 0
    """Lower is preferred. Used to order the fallback chain."""
    timeout_s: float = DEFAULT_API_TIMEOUT_S
    max_concurrency: int = 64
    #: Serverless/cold-start backends carry extra latency risk.
    cold_start_risk: float = 0.0
    #: Self-hosted backends have a marginal cost near zero but finite capacity.
    cost_factor: float = 1.0
    """Multiplier on the model's list price. Self-hosted ≈ 0.1–0.3 of API price
    once you're above the GPU duty-cycle break-even, but see `capacity`."""
    capacity_rps: float | None = None
    """If set, the backend sheds load past this rate instead of queueing."""

    def supports(self, model_id: str) -> bool:
        return not self.models or model_id in self.models


@dataclass
class BackendHealth:
    """Live signals. Updated by the execution layer from real responses."""

    backend_id: str
    in_flight: int = 0
    successes: int = 0
    failures: int = 0
    consecutive_failures: int = 0
    ewma_latency_ms: float = 0.0
    p99_latency_ms: float = 0.0
    open_until: float = 0.0
    """Circuit open until this monotonic timestamp."""
    half_open_probes: int = 0
    rate_limited_until: float = 0.0
    last_error: str | None = None
    _latency_alpha: float = 0.2

    @property
    def error_rate(self) -> float:
        total = self.successes + self.failures
        return self.failures / total if total else 0.0

    def is_available(self, now: float | None = None) -> bool:
        now = now if now is not None else time.monotonic()
        if self.rate_limited_until and now < self.rate_limited_until:
            return False
        return now >= self.open_until

    def is_half_open(self, now: float | None = None) -> bool:
        now = now if now is not None else time.monotonic()
        return 0 < self.open_until <= now

    def record_success(self, latency_ms: float, now: float | None = None) -> None:
        now = now if now is not None else time.monotonic()
        self.successes += 1
        self.consecutive_failures = 0
        self.ewma_latency_ms = (
            (1 - self._latency_alpha) * self.ewma_latency_ms
            + self._latency_alpha * latency_ms
        )
        self.p99_latency_ms = max(self.p99_latency_ms * 0.95, latency_ms)
        self.open_until = 0.0
        self.half_open_probes = 0
        self.last_error = None

    def record_failure(self, error: str, retry_after_s: float | None = None,
                       now: float | None = None) -> None:
        now = now if now is not None else time.monotonic()
        self.failures += 1
        self.consecutive_failures += 1
        self.last_error = error
        if retry_after_s:
            self.rate_limited_until = now + retry_after_s
        if self.consecutive_failures >= CIRCUIT_FAILURE_THRESHOLD:
            backoff = min(
                CIRCUIT_BASE_BACKOFF_S * (2 ** (self.consecutive_failures
                                                - CIRCUIT_FAILURE_THRESHOLD)),
                CIRCUIT_MAX_BACKOFF_S,
            )
            self.open_until = now + backoff
            # A failure while half-open re-opens the circuit; any in-flight
            # canary allowance is revoked with it.
            self.half_open_probes = 0

    def saturating(self, spec: BackendSpec) -> bool:
        """True when the backend should be shed rather than queued."""
        if self.in_flight >= spec.max_concurrency:
            return True
        if spec.capacity_rps is not None and self.in_flight > spec.capacity_rps:
            return True
        return False


CIRCUIT_FAILURE_THRESHOLD = 5
CIRCUIT_BASE_BACKOFF_S = 1.0
CIRCUIT_MAX_BACKOFF_S = 60.0

#: How many probe requests may be in flight against a HALF_OPEN backend. Full
#: recovery is gated behind a cheap canary: one probe, and only after it
#: succeeds does the circuit close. Never probe by calling the provider for
#: health's sake — the canary is a real request that was going out anyway.
HALF_OPEN_PROBE_LIMIT = 1


class BackendRegistry:
    def __init__(self, backends: Sequence[BackendSpec] = ()) -> None:
        self._specs: dict[str, BackendSpec] = {}
        self._health: dict[str, BackendHealth] = {}
        for b in backends:
            self.register(b)

    def register(self, spec: BackendSpec) -> None:
        self._specs[spec.backend_id] = spec
        self._health.setdefault(spec.backend_id, BackendHealth(spec.backend_id))

    def unregister(self, backend_id: str) -> None:
        self._specs.pop(backend_id, None)
        self._health.pop(backend_id, None)

    def spec(self, backend_id: str) -> BackendSpec | None:
        return self._specs.get(backend_id)

    def health(self, backend_id: str) -> BackendHealth:
        h = self._health.get(backend_id)
        if h is None:
            h = BackendHealth(backend_id)
            self._health[backend_id] = h
        return h

    def candidates_for(self, model_id: str) -> tuple[BackendSpec, ...]:
        return tuple(b for b in self._specs.values() if b.supports(model_id))

    def available_for(self, model_id: str, now: float | None = None) -> tuple[BackendSpec, ...]:
        """Healthy, non-saturated backends that serve this model, by priority.

        A HALF_OPEN backend is admitted only while a canary slot is free: full
        recovery waits for one real request to succeed.
        """
        now = now if now is not None else time.monotonic()
        out = []
        for b in self.candidates_for(model_id):
            h = self.health(b.backend_id)
            if not h.is_available(now) or h.saturating(b):
                continue
            if h.is_half_open(now) and h.half_open_probes >= HALF_OPEN_PROBE_LIMIT:
                continue
            out.append(b)
        return tuple(sorted(out, key=lambda b: b.priority))

    def all_available(self, now: float | None = None) -> tuple[BackendSpec, ...]:
        now = now if now is not None else time.monotonic()
        return tuple(
            b for b in self._specs.values() if self.health(b.backend_id).is_available(now)
        )

    def ids(self) -> tuple[str, ...]:
        return tuple(self._specs)

    def stats(self) -> Mapping[str, Mapping[str, float]]:
        return {
            bid: {
                "in_flight": float(h.in_flight),
                "error_rate": h.error_rate,
                "ewma_latency_ms": h.ewma_latency_ms,
                "consecutive_failures": float(h.consecutive_failures),
                "circuit_open": 1.0 if h.open_until else 0.0,
            }
            for bid, h in self._health.items()
        }

    def __len__(self) -> int:
        return len(self._specs)


__all__ = [
    "DEFAULT_LOCAL_TIMEOUT_S",
    "DEFAULT_API_TIMEOUT_S",
    "CIRCUIT_FAILURE_THRESHOLD",
    "CIRCUIT_BASE_BACKOFF_S",
    "CIRCUIT_MAX_BACKOFF_S",
    "BackendSpec",
    "BackendHealth",
    "BackendRegistry",
]

File: assets/engine/backends.js
/**
 * Backend registry, live health signals, the circuit breaker.
 * Axis 2: WHERE the model runs. The router scores (model, backend) PAIRS.
 * Local timeouts are aggressive on purpose: a saturated GPU that cascades to
 * cloud at 100x cost is the trap.
 */
import { nowMonotonic } from "./clock.js";
import { DeploymentClass } from "./types.js";
export const DEFAULT_LOCAL_TIMEOUT_S = 8.0;
export const DEFAULT_API_TIMEOUT_S = 60.0;
export const CIRCUIT_FAILURE_THRESHOLD = 5;
export const CIRCUIT_BASE_BACKOFF_S = 1.0;
export const CIRCUIT_MAX_BACKOFF_S = 60.0;
/** Probe requests allowed in flight against a HALF_OPEN backend. */
export const HALF_OPEN_PROBE_LIMIT = 1;
export class BackendSpec {
    backendId;
    deployment;
    models;
    baseUrl;
    priority;
    timeoutS;
    maxConcurrency;
    coldStartRisk;
    costFactor;
    capacityRps;
    constructor(backendIdOrOpts, maybeOpts = {}) {
        const opts = typeof backendIdOrOpts === "string" ? maybeOpts : backendIdOrOpts;
        const backendId = typeof backendIdOrOpts === "string"
            ? backendIdOrOpts
            : backendIdOrOpts.backendId;
        if (!backendId)
            throw new Error("BackendSpec requires a backendId");
        this.backendId = backendId;
        this.deployment = opts.deployment ?? DeploymentClass.Api;
        this.models = opts.models ?? [];
        this.baseUrl = opts.baseUrl ?? null;
        this.priority = opts.priority ?? 0;
        this.timeoutS = opts.timeoutS ?? DEFAULT_API_TIMEOUT_S;
        this.maxConcurrency = opts.maxConcurrency ?? 64;
        this.coldStartRisk = opts.coldStartRisk ?? 0.0;
        this.costFactor = opts.costFactor ?? 1.0;
        this.capacityRps = opts.capacityRps ?? null;
    }
    supports(modelId) {
        return this.models.length === 0 || this.models.includes(modelId);
    }
}
export class BackendHealth {
    backendId;
    inFlight = 0;
    successes = 0;
    failures = 0;
    consecutiveFailures = 0;
    ewmaLatencyMs = 0.0;
    p99LatencyMs = 0.0;
    /** Circuit open until this monotonic timestamp. */
    openUntil = 0.0;
    halfOpenProbes = 0;
    rateLimitedUntil = 0.0;
    lastError = null;
    latencyAlpha = 0.2;
    constructor(backendId) {
        this.backendId = backendId;
    }
    get errorRate() {
        const total = this.successes + this.failures;
        return total === 0 ? 0.0 : this.failures / total;
    }
    isAvailable(now = nowMonotonic()) {
        if (this.rateLimitedUntil && now < this.rateLimitedUntil)
            return false;
        return now >= this.openUntil;
    }
    isHalfOpen(now = nowMonotonic()) {
        return this.openUntil > 0 && this.openUntil <= now;
    }
    recordSuccess(latencyMs, now = nowMonotonic()) {
        this.successes += 1;
        this.consecutiveFailures = 0;
        this.ewmaLatencyMs =
            (1 - this.latencyAlpha) * this.ewmaLatencyMs + this.latencyAlpha * latencyMs;
        this.p99LatencyMs = Math.max(this.p99LatencyMs * 0.95, latencyMs);
        this.openUntil = 0.0;
        this.halfOpenProbes = 0;
        this.lastError = null;
    }
    recordFailure(error, retryAfterS = null, now = nowMonotonic()) {
        this.failures += 1;
        this.consecutiveFailures += 1;
        this.lastError = error;
        if (retryAfterS)
            this.rateLimitedUntil = now + retryAfterS;
        if (this.consecutiveFailures >= CIRCUIT_FAILURE_THRESHOLD) {
            const backoff = Math.min(CIRCUIT_BASE_BACKOFF_S *
                2 ** (this.consecutiveFailures - CIRCUIT_FAILURE_THRESHOLD), CIRCUIT_MAX_BACKOFF_S);
            this.openUntil = now + backoff;
            // A failure while half-open re-opens the circuit and revokes any
            // in-flight canary allowance.
            this.halfOpenProbes = 0;
        }
    }
    saturating(spec) {
        if (this.inFlight >= spec.maxConcurrency)
            return true;
        if (spec.capacityRps !== null && this.inFlight > spec.capacityRps)
            return true;
        return false;
    }
}
export class BackendRegistry {
    specs = new Map();
    healthMap = new Map();
    constructor(backends = []) {
        for (const b of backends)
            this.register(b);
    }
    register(spec) {
        this.specs.set(spec.backendId, spec);
        if (!this.healthMap.has(spec.backendId)) {
            this.healthMap.set(spec.backendId, new BackendHealth(spec.backendId));
        }
    }
    unregister(backendId) {
        this.specs.delete(backendId);
        this.healthMap.delete(backendId);
    }
    spec(backendId) {
        return this.specs.get(backendId);
    }
    health(backendId) {
        let h = this.healthMap.get(backendId);
        if (!h) {
            h = new BackendHealth(backendId);
            this.healthMap.set(backendId, h);
        }
        return h;
    }
    /** Read-only view for state snapshots (see store.ts). */
    allHealth() {
        return this.healthMap.values();
    }
    candidatesFor(modelId) {
        return [...this.specs.values()].filter((b) => b.supports(modelId));
    }
    /** Healthy, non-saturated backends serving this model, by priority. */
    availableFor(modelId, now = nowMonotonic()) {
        const out = this.candidatesFor(modelId).filter((b) => {
            const h = this.health(b.backendId);
            if (!h.isAvailable(now) || h.saturating(b))
                return false;
            // HALF_OPEN: admit only while a canary slot is free.
            if (h.isHalfOpen(now) && h.halfOpenProbes >= HALF_OPEN_PROBE_LIMIT)
                return false;
            return true;
        });
        return out.sort((a, b) => a.priority - b.priority);
    }
    ids() {
        return [...this.specs.keys()];
    }
    get size() {
        return this.specs.size;
    }
}
